### ISSUE TITLE
Type narrowing support for expressions that contain a subscript

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#1494]
**Describe the bug**
If you have a variable of type `Tuple[int, Optional[str]]` and you test that the second component is not `None`, pyright says that you can't assume the type of that component is not `None`

**To Reproduce**
```python
from typing import Tuple, Optional


def myfunc(x: Tuple[int, Optional[str]]) -> str:
    y: str = ""
    if x[1] is not None:
        y = x[1]
    return y


x = (1, "abc")

print(myfunc(x))
```
Reports
```
Expression of type "str | None" cannot be assigned to declared type "str"
  Type "str | None" cannot be assigned to type "str"
    Type "None" cannot be assigned to type "str"
```
corresponding to the line `y = x[1]`

**Expected behavior**
No error message

**VS Code extension or command-line**
pyright 1.1.110  command line 

**Additional context**
`mypy` does not flag this as an error.

**Comment by erictraut** [comment]
This is intentional. Type narrowing is not supported for expressions that include subscripts. The object being subscripted can change in too many ways, so it's not safe to assume that `x[y]` is going to evaluate to the same value (and therefore the same type) if executed multiple times within a scope. If `y` is a literal value, it becomes less dangerous, but it's still a problem if the container object is mutated in ways that the type checker can't understand (e.g. by calling `pop` on a list object).

AFAIK, none of the other Python type checkers perform type narrowing on expressions that contain subscripts either, for the same reasons I mentioned above.

**Comment by erictraut** [comment]
Here are a couple of simple workarounds:

```python
def myfunc(x: Tuple[int, Optional[str]]) -> str:
    y: str = ""
    if (a := x[1]) is not None:
        y = a
    return y
```

or

```python
def myfunc(x: Tuple[int, Optional[str]]) -> str:
    y: str = ""
    a = x[1]
    if a is not None:
        y = a
    return y
```

**Comment by Dr-Irv** [comment]
> AFAIK, none of the other Python type checkers perform type narrowing on expressions that contain subscripts either, for the same reasons I mentioned above.

The code I provided passes through `mypy` without any error reports.

> The object being subscripted can change in too many ways, so it's not safe to assume that x[y] is going to evaluate to the same value (and therefore the same type) if executed multiple times within a scope.

Yes, but from a type perspective, on a tuple, if the tuple is declared as `Tuple(int, Optional[str])`, then you know that `x[1]` has type `Optional[str]`, and if the test `if x[1] is not None` is done, then you know the type of `x[1]` is `str`

I agree with your assessment on general subscripting, but I think declared tuples are different because the type of each element is known.

**Comment by erictraut** [comment]
Interesting, it looks like mypy added type narrowing support for subscripted expressions. Earlier versions did not support this.

The issue isn't that types are known for elements of a tuple. (After all, the types of elements in a list are also known.) The important distinction is that tuples are immutable, and `x[1]` is guaranteed to to return the same value if it is evaluated multiple times within the same code flow. The same cannot be said for mutable containers list lists.

I don't consider the current behavior a bug. Type narrowing is currently provided only for [this set](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#type-guards) of expression types. I will take this as a feature request. It's a pretty significant one though. I need to think more about the tradeoffs before I commit to implementing some subset of support for type narrowing of expressions with subscripts.

**Comment by Dr-Irv** [comment]
So in our code (this is not pandas stuff, BTW), we have something like this:
```python
def myfunc2(x: Optional[Tuple[int, Optional[str]]]) -> str:
    if x is None or x[1] is None:
        raise Exception("cannot pass None")
    y: str = x[1]
    return y
```

`pyright` complains again about the assignment to `y` but at this point, you know it must be `str` .  I can't figure out a workaround along the pattern you suggested above without the code looking very clumsy.  But maybe I'm missing something.

**Comment by erictraut** [comment]
Without conditional type narrowing, the expression `x[1]` will always evaluate to type `Optional[str]` even within a conditional block. Type narrowing is what allows a type checker to say "because of this conditional check, the type is no longer `Optional[str]` but is instead just `str`".

The workaround is the same as what I suggested above, assuming you are using a version of Python that supports the walrus operator.

```python
def myfunc2(x: Optional[Tuple[int, Optional[str]]]) -> str:
    if x is None or (a := x[1]) is None:
        raise Exception("cannot pass None")
    y: str = a
    return y
```

**Comment by erictraut** [comment]
I've added type narrowing support for index expressions where the subscript is specified as a numeric (integral) literal. This covers common cases (including the sample at the top of this enhancement request), and it is relatively safe. This is the same limited set of cases that mypy appears to support. It's also the same subset of cases supported by the TypeScript compiler.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.113, which I just published. It will also be included in the next release of Pylance.
